include/proxy-wasm/context.h

```diff
@@ -0,0 +1,306 @@
+  ContextBase(WasmBase *wasm, std::shared_ptr<PluginBase> plugin); // Root Context.
+  ContextBase(WasmBase *wasm, uint32_t root_context_id,
+              std::shared_ptr<PluginBase> plugin); // Stream context.
+  virtual ~ContextBase();
+
+  WasmBase *wasm() const { return wasm_; }
+  uint32_t id() const { return id_; }
+  // The VM Context used for calling "malloc" has an id_ == 0.
+  bool isVmContext() { return id_ == 0; }
+  // Root Contexts have the VM Context as a parent.
+  bool isRootContext() { return root_context_id_ == 0; }
+  ContextBase *root_context() { return root_context_; }
+  string_view root_id() const { return plugin_->root_id_; }
+  string_view log_prefix() const { return plugin_->log_prefix(); }
+  WasmVm *wasmVm() const;
+
+  /**
+   * Calls into the VM.
+   * These are implemented by the proxy-independent host code. They are virtual to support some
+   * types of testing.
+   */
+
+  // Context
+  void onCreate(uint32_t parent_context_id) override;
+  bool onDone() override;
+  void onDelete() override;
+
+  // Root
+  bool onStart(std::shared_ptr<PluginBase> plugin) override;
+  bool onConfigure(std::shared_ptr<PluginBase> plugin) override;
+  void onHttpCallResponse(HttpCallToken token) override;
+  void onQueueReady(uint32_t SharedQueueDequeueToken) override;
+  void onGrpcReceiveInitialMetadata(GrpcToken token) override;
+  void onGrpcReceive(GrpcToken token) override;
+  void onGrpcReceiveTrailingMetadata(GrpcToken token) override;
+  void onGrpcClose(GrpcToken token, const GrpcStatus &status, const string_view message) override;
+
+  // Stream (e.g. HTTP, Network)
+  void onFinalized() override;
+  const BufferInterface *getBuffer(WasmBufferType /* type */) override {
+    unimplemented();
+    return nullptr;
+  }
+  bool end_of_stream(WasmStreamType /* type */) override {
+    unimplemented();
+    return true;
+  }
+  WasmResult continueStream(WasmStreamType /* type */) override { return unimplemented(); }
+
+  // HTTP
+  ProxyAction onRequestHeaders() override;
+  ProxyAction onRequestBody() override;
+  ProxyAction onRequestTrailers() override;
+  ProxyAction onRequestMetadata() override;
+  ProxyAction onResponseHeaders() override;
+  ProxyAction onResponseBody() override;
+  ProxyAction onResponseTrailers() override;
+  ProxyAction onResponseMetadata() override;
+  WasmResult sendLocalResponse(uint64_t /* response_code */, string_view /* body */,
+                               Pairs /* additional_headers */, uint64_t /* grpc_status */,
+                               string_view /* details */) override {
+    return unimplemented();
+  }
+
+  // Network
+  ProxyAction onNetworkNewConnection() override;
+  ProxyAction onDownstreamData() override;
+  ProxyAction onUpstreamData() override;
+  using CloseType = NetworkInterface::CloseType;
+  void onDownstreamConnectionClose(CloseType close_type) override;
+  void onUpstreamConnectionClose(CloseType close_type) override;
+
+  // General
```

Comment:
suggest outlining all these less-than-trivial functions into a .cc file.

Reply:
These have full implementations.  See the master branch.   This is just a review of the header.   We have changed our review plan and will be reviewing on the envoy-wasm repo after this PR.